esrally/telemetry.py

```diff
@@ -743,6 +743,181 @@ def sample(self):
+                    raise exceptions.SystemSetupError(
+                        f"The telemetry parameter 'transform-stats-transforms' must be a JSON Object with keys "
+                        f"matching the cluster names [{','.join(sorted(clients.keys()))}] specified in --target-hosts "
+                        f"but it had [{cluster_name}].")
+            self.specified_cluster_names = self.transforms_per_cluster.keys()
+
+        self.metrics_store = metrics_store
+        self.samplers = []
+
+    def on_benchmark_start(self):
+        recorder = []
+        for cluster_name in self.specified_cluster_names:
+            recorder = TransformStatsRecorder(cluster_name, self.clients[cluster_name], self.metrics_store,
+                                              self.sample_interval,
+                                              self.transforms_per_cluster[
+                                                  cluster_name] if self.transforms_per_cluster else None)
+            sampler = SamplerThread(recorder)
+            self.samplers.append(sampler)
+            sampler.setDaemon(True)
+            # we don't require starting recorders precisely at the same time
+            sampler.start()
+
+    def on_benchmark_stop(self):
+        if self.samplers:
+            for sampler in self.samplers:
+                sampler.finish()
+                # record the final stats
+                sampler.recorder.record_final()
+
+
+class TransformStatsRecorder:
+    """
+    Collects and pushes Transform stats for the specified cluster to the metric store.
+    """
+
+    def __init__(self, cluster_name, client, metrics_store, sample_interval, transforms=None):
+        """
+        :param cluster_name: The cluster_name that the client connects to, as specified in target.hosts.
+        :param client: The Elasticsearch client for this cluster.
+        :param metrics_store: The configured metrics store we write to.
+        :param sample_interval: integer controlling the interval, in seconds, between collecting samples.
+        :param transforms: optional list of transforms to filter results from.
+        """
+
+        self.cluster_name = cluster_name
+        self.client = client
+        self.metrics_store = metrics_store
+        self.sample_interval = sample_interval
+        self.transforms = transforms
+        self.logger = logging.getLogger(__name__)
+
+        self.logger.info("transform stats recorder")
+
+    def __str__(self):
+        return "transform stats"
+
+    def record(self):
+        """
+        Collect Transform stats for transforms (optionally) specified in telemetry parameters and push to metrics store.
+        """
+
+        self._record(metrics.SampleType.Normal)
+
+    def record_final(self):
+        """
+        Collect final Transform stats for transforms (optionally) specified in telemetry parameters and push to metrics store.
+        """
+
+        self._record(metrics.SampleType.Final)
```

Comment:
^ I added `SampleType.Final` in order to query only for the last, final record in the end result.

Intermediate results are indexed with `SampleType.Normal`

I am not sure if that's an abusive usage of `SampleType`

Reply:
I think we can avoid introducing a new sample type for this. Given that the final sample is a total summary at the end we could prefix each metric name with `total_` in the telemetry device and then only query for that in the stats calculator?